File: Arrays & Strings/URLify.java
// Cracking the Coding Interview - pp. 90 - q 1.3

public class URLify {
    
    public static String solOne(String str) {

        // OBJECTIVE: Replace all whitespace with "%20" and return string

        // Remove duplicate whitespaces with replaceAll()
        // Then, remove leading and trailing whitespaces with trim()
        String trimmed = str.replaceAll("\\s+", " ").trim();

        // Traverse string
        StringBuilder ans = new StringBuilder();
        for (char c: trimmed.toCharArray()) {

            // Replace whitespace with "%20"
            if (c == ' ') {
                ans.append("%20");
            }
            else {
                ans.append(c);
            }
        }

        // Return StringBuilder as a string
        return ans.toString();
    }

    public static void main(String args[]) {
        System.out.println(solOne("   John Smith   "));
    }
}

File: Arrays & Strings/URLify.py
# Cracking the Coding Interview - pp. 90 - q 1.3

class URLify:

    def solOne(self, str1):

        # OBJECTIVE: Replace all whitespaces with "%20" inside string

        # Replace all duplicate whitespace with a singular one
        str1 = str1.replace("  ", " ")

        # Remove leading and trailing whitespaces
        str1 = str1.strip()

        # Convert string to a list
        str1 = list(str1)

        # Traverse string
        for i in range(len(str1)):

            if str1[i] == " ":
                str1[i] = "%20"

        # Return list as string
        return "".join(i for i in str1)

def main():

    ans = URLify()
    print(ans.solOne("  John Smith  "))

main()